docs/2019-03-28-All-Path-Reachability-Proofs.md

```diff
@@ -0,0 +1,253 @@
+Proving All-Path Reachability Claims
+====================================
+
+This document details All-Path Reachability without solving the
+most-general-unifier (MGU) problem.
+MGU will be detailed in a separate document.
+
+Background
+----------
+
+Recommended reading: [proving One-Path Reachability
+claims](2018-11-08-One-Path-Reachability-Proofs.md).
+
+Similarly to the One-Path Reachability document, we assume all pattern variables
+used in this document to be _extended function-like patterns_, that is patterns
+which can be written as `t ∧ p` where the interpretation of `t` contains at most
+one element and `p` is a predicate.
+
+_Extended constructor patterns_ will be those extended function-like patterns
+for which `t` is a functional term, composed out of constructor-like symbols
+and variables.
+
+
+Definitions
+-----------
+
+### Weak always
+
+Given a formula `φ`, let `[w]φ` denote the formula “weak always” `φ`,
+defined by:
+
+```
+[w]φ := νX.φ ∨ (○X ∧ •⊤)
+```
+
+one consequence of the above is that `[w]φ = φ ∨ (○[w]φ ∧ •⊤)`
+
+
+Problem Description
+-------------------
+
+Given a set of all-path reachability claims, of the form `∀x.φ(x) → [w]∃z.ψ(x,z)`,
+we are trying to prove all of them together.
+
+
+Algorithms
+----------
+(detailed in the next section)
+
+
+### Algorithm `proveAllPath`
```

Comment:
It looks like the algorithm derivates from the classic all-path reachability logic prover, where circularities (claims) are regarded as semantic axioms, and (Circurlarity) rule is forbidden. Is it intended?

Reply:
> Since the algorithm may not terminate, it's good to have a bound of the number of while iterations. If the bound is reached, then it returns `Unproved`.

@dlucanu : I added a note about this close to the description of the algorithm.